central/rbac/service/list_subjects.go

```diff
@@ -137,28 +140,33 @@ func NewSubjectSearcher(k8sRoleBindingDatastore datastore.DataStore) *SubjectSea
 
 // Search implements the searcher interface
 func (s *SubjectSearcher) Search(ctx context.Context, q *v1.Query) ([]search.Result, error) {
-	subjectQuery, _ := search.FilterQueryWithMap(q, mapping.OptionsMap)
-	pred, err := subjectFactory.GeneratePredicate(subjectQuery)
+	roleBindingQuery, _ := search.FilterQueryWithMap(q, schema.RoleBindingsSchema.OptionsMap)
+	roleBindingPred, err := roleBindingPredFactory.GeneratePredicate(roleBindingQuery)
 	if err != nil {
 		return nil, err
 	}
 
-	bindings, err := s.k8sRoleBindingDatastore.SearchRawRoleBindings(ctx, q)
+	subjectQuery, _ := search.FilterQueryWithMap(q, mapping.OptionsMap)
+	subjectPred, err := subjectFactory.GeneratePredicate(subjectQuery)
 	if err != nil {
 		return nil, err
 	}
-	// Subject return only users and groups, there is a separate resolver for service accounts.
-	subjects := k8srbac.GetAllSubjects(bindings, storage.SubjectKind_USER, storage.SubjectKind_GROUP)
-	// Sort first then evaluate to not run evaluation twice.
-	// Sorting should be cheaper than reflect based evaluation
-	if err := sortSubjects(q, subjects); err != nil {
+
+	bindings, err := s.k8sRoleBindingDatastore.SearchRawRoleBindings(ctx, roleBindingQuery)
+	if err != nil {
 		return nil, err
 	}
 
 	var results []search.Result
-	for _, subject := range subjects {
-		if result, match := pred.Evaluate(subject); match {
-			results = append(results, *result)
+	for _, binding := range bindings {
+		if bindingResult, match := roleBindingPred.Evaluate(binding); match {
+			subjects := k8srbac.GetAllSubjects([]*storage.K8SRoleBinding{binding}, storage.SubjectKind_USER, storage.SubjectKind_GROUP)
```

Comment:
Cannot move this check into the query passed to `SearchRawRoleBindings` because if `Search` is called with an unsupported field like `Deployment:dep`, then the search should return 0 results. If SubjectKind conjunction is added to q, `SearchRawRoleBindings` will find role bindings and pred evaluation loop will include matches.

Also, not directly adding SubjectKind conjunction to `roleBindingPred` and `subjectPred` because we do not want matches to have `subject.kind : <values>` when query did not ask for it.